### PR TITLE
add support for durable logging

### PR DESCRIPTION
**Pull request by MiguelAlho**
- Support for durable logging is based on the implementation of the Seq sink.
- ControlledSwitch, ExponentialBackoff and PortableTimes have been imported directly (except for the namespace change.
- LogglySink uses the same strategy of Seq Sink to determine if the client is durable or not.
- HttpLogShipper uses the LogglyClient from the loggly-csharp lib instead of the http client directly, and can use the bulk api correctly.
- DurableSink uses rolling file sink (like Seq's durableSink) to store events, but serializes them as logglyEvents instead of serilog's logevent. This avoid problems reading the serialized events from the file.
- Event conversion is in LogEventConverter class for reuse

### CODE REVIEW

File: src/Serilog.Sinks.Loggly/LoggerConfigurationLogglyExtensions.cs [hunk 1]
```diff
@@ -32,22 +35,65 @@ public static class LoggerConfigurationLogglyExtensions
...
-                restrictedToMinimumLevel);
+            ILogEventSink sink;
+
+            if (bufferBaseFilename == null)
+            {
+                sink = new LogglySink(formatProvider, batchPostingLimit, defaultedPeriod);
+            }
+            else
+            {
+#if DURABLE
```
Comment (nblumhardt) [inline src/Serilog.Sinks.Loggly/LoggerConfigurationLogglyExtensions.cs]: All of the targeted frameworks define `DURABLE`, so the `#if` blocks throughout are not necessary.

**Comment by nblumhardt** [comment]
Thanks for the PR, @MiguelAlho.

My only real concern with this is that it's a lot of code to support, while there haven't been a lot of long-term maintainers working on this sink.

If I can ask... what's the time-frame of the project you're using this code on? Are you likely to be able to troubleshoot during the first year or so of this being in the wild?

Cheers!

File: sample/sampleDurableLogger/Program.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+using Serilog.Sinks.RollingFileAlternate;
+
+namespace sampleDurableLogger
+{
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            SetupLogglyConfiguration();
+            var logger = CreateLoggerFactory(@"c:\test\").CreateLogger();
```
Comment (MiguelAlho) [inline sample/sampleDurableLogger/Program.cs]: Yes. My use-case is more along the use of the static members, but for the purpose of the sample, the using should be present.

File: sample/sampleDurableLogger/Program.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            Console.ReadLine();
+
+            logger.Information((Exception)null, "third test message - app started");
+            logger.Warning((Exception)null, "third test message with {@data}", new { p1 = "sample3", p2 = DateTime.Now });
+
+            Console.WriteLine("back online messages written. Check loggly and files for data. wait a minute or so before reconnecting. Press Enter to temrinate");
+            Console.ReadLine();
+        }
+
+        public static LoggerConfiguration CreateLoggerFactory(string logFilePath)
```
Comment (MiguelAlho) [inline sample/sampleDurableLogger/Program.cs]: I'll change it to CreateLogger and return an ILogger directly. This was a copy / paste mistake from my implementation case-

**Comment by MiguelAlho** [comment]
@nblumhardt , Thanks for the detailed review. I've updated the code in the PR to reflect the changes you've mentioned and also fixed a few more types, modifiers and namespaces.

My team has been integrating structured logging into our legacy (and any new) apps with Serilog, and we are opting in for Loggly for remote logging since it has pretty much all we need and won't require much infrastructure maintenance on our side. We are committed to it, and will be using it a lot here so I don't see any problem in us help maintaining this package (and any others we may need). The only reason to not use it would be if we some how grew out of what we require from Loggly, but for now it seems to cover all of our usage scenarios. 

There is some functional coupling to the Seq sink, since this integration was based on it. I'll follow that repo and try to see if any changes there should be ported over. Also, I might make sense to consider extracting common parts to a Durable sink (abstractions?) if more remote sinks should support this capability.

**Comment by nblumhardt** [comment]
Cool, thanks for the follow-up Miguel. I'm on board with merging this one 👍

Extracting the common "durable" functionality does seem like a worthwhile move, but there are some challenges in getting the right design for it (just extracting the common parts into a library would deal with code reuse, but without the right level of abstraction I think it could be hard to make improvements to the shared code without downstream breakages). Just a design task, but would take some time :-)

I'll add you to the "reviewers" team on this repository, thanks for the help!

Nick

**Comment by nblumhardt** [comment]
BTW, it would be great to have a note/sample in the README for this feature, if anyone has a chance.

**Pull request merged**